### PR TITLE
(TK-160) Initial scaffolding for conf parser tests

### PR DESCRIPTION
**Pull request by cprice404**
This commit ports over the first few ConfParser tests, and fixes
a few bugs to get them passing.

### CODE REVIEW

File: lib/hocon/impl/parseable.rb [hunk 1]
```diff
@@ -101,7 +101,11 @@ def parse
     self.class.force_parsed_to_object(parse_value(options))
   end
 
-  def parse_value(base_options)
+  def parse_value(base_options = nil)
+    if base_options.nil?
+      base_options = options
+    end
+
```
Comment (KevinCorcoran) [inline lib/hocon/impl/parseable.rb]: Could you change the default value of this param to just be `options`?  So, `def parse_value(base_options = options)`.  It's not exactly the same, but maybe sufficient?

File: spec/unit/typesafe/config/conf_parser_spec.rb [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+      Hocon::ConfigResolveOptions.no_system)
+  else
+    tree
+  end
+end
+
+
+describe "Config Parser" do
+   context "invalid_conf_throws" do
+     TestUtils.whitespace_variations(TestUtils::InvalidJsonInvalidConf, false).each do |t|
```
Comment (KevinCorcoran) [inline spec/unit/typesafe/config/conf_parser_spec.rb]: This would be a little better if `t` was renamed to `invalid` to match upstream.

**Comment by KevinCorcoran** [comment]
:+1: 

I skimmed `TestUtils` because it's insane, but all looks good otherwise.  I'm fine with this being merged as-is.

@jpinsonault you're up.  :)

File: lib/hocon/impl/parseable.rb (hunk 1 above)
Comment (cprice404) [inline lib/hocon/impl/parseable.rb]: I can never remember the language-specific rules for when those defaults get evaluated.  In some languages the default only gets evaluated once, when the code is loaded, and that can cause weird behaviors, so I generally try to just avoid putting method calls as defaults in the args list just to be safe.

File: spec/unit/typesafe/config/conf_parser_spec.rb (hunk 2 above)
Comment (cprice404) [inline spec/unit/typesafe/config/conf_parser_spec.rb]: I can do that.

File: lib/hocon/impl/parseable.rb (hunk 1 above)
Comment (KevinCorcoran) [inline lib/hocon/impl/parseable.rb]: oh, i didn't realize `options` was a method.  good call on avoiding that.

**Comment by cprice404** [comment]
Changed `t` to `invalid`.

File: spec/test_utils.rb [hunk 3]
```diff
@@ -123,7 +315,7 @@ def hash
   ##################
   # Path Functions
   ##################
-  def TestUtils.path(*elements)
+  def self.path(*elements)
```
Comment (jpinsonault) [inline spec/test_utils.rb]: def self.* makes this much easier to read, good call

File: spec/unit/typesafe/config/conf_parser_spec.rb [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+   context "invalid_conf_throws" do
+     TestUtils.whitespace_variations(TestUtils::InvalidJsonInvalidConf, false).each do |invalid|
+       TestUtils.add_offending_json_to_exception("config", invalid.test) {
+         TestUtils.intercept(Hocon::ConfigError) {
+           parse(invalid.test)
+         }
+       }
+     end
+   end
+end
```
Comment (jpinsonault) [inline spec/unit/typesafe/config/conf_parser_spec.rb]: Do we feel strongly about ending files with newlines?

**Comment by jpinsonault** [comment]
One minor question, otherwise this looks good! :+1:

File: spec/unit/typesafe/config/conf_parser_spec.rb (hunk 4 above)
Comment (cprice404) [inline spec/unit/typesafe/config/conf_parser_spec.rb]: jeff does.  :)

**Comment by fpringvaldsen** [comment]
I'm also :+1:

**Comment by cprice404** [comment]
added missing newline

**Pull request merged**